services/ingestion/engine.go

```diff
@@ -163,14 +186,36 @@ func (e *Engine) processEvents(events *models.CadenceEvents) error {
 		Int("cadence-event-length", events.Length()).
 		Msg("received new cadence evm events")
 
-	batch := e.store.NewBatch()
-	defer func(batch *pebbleDB.Batch) {
-		err := batch.Close()
-		if err != nil {
-			e.log.Fatal().Err(err).Msg("failed to close batch")
+	err := e.withBatch(
+		func(batch *pebbleDB.Batch) error {
+			return e.indexEvents(events, batch)
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to index events for cadence block %d: %w", events.CadenceHeight(), err)
+	}
+
+	e.collector.CadenceHeightIndexed(events.CadenceHeight())
+
+	if events.Empty() {
+		return nil // nothing else to do this was heartbeat event with not event payloads
+	}
+
+	// emit block event and logs, only after we successfully commit the data
+	e.blocksPublisher.Publish(events.Block())
```

Comment:
Who depends on this publisher? They need to be aware if a crash happens before this call and after the commit, then the subscribe will miss the data.

Reply:
This only notifies the streaming api, that a new block is  available.

If a crash happens after commit and before this (no errrors are expected but there could be a panic or outage), the clients on the streaming api will be disconnected, and they will need to handle getting missing blocks and re-subscribing once the GW is back up. This should be already standard procedure if a disconnect happens.